File: dsa/QuickSort.java
public class QuickSort {
    public static int partition(int[] arr, int low, int high) {
        int pivot = arr[high];
        int i = low  - 1;
        
        for(int j = low; j < high; j++) {
            if(arr[j] < pivot) {
                int temp = arr[++i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }

        int temp = arr[++i];
        arr[i] = arr[high];
        arr[high] = temp;

        return i;
    }

    public static void quickSort(int[] arr, int low, int high) {
        if (low < high) {
            int pidx = partition(arr, low, high);
            quickSort(arr, low, pidx-1);
            quickSort(arr, pidx+1, high);
        }
    }

    public static void main(String[] args) {
        int[] arr = {2, 6, 3, 9, 5, 7, 8};

        quickSort(arr, 0, arr.length-1);

        for(int i : arr) System.out.print(i + " ");
        System.out.println();
    }
}

File: dsa/quickSort.cpp
#include <iostream>
using namespace std;

int partition(int *arr, int s, int e)
{
    int pivot = arr[s], count = 0;

    for (int i = s + 1; i <= e; i++)
    {
        if (arr[i] <= pivot)
            count++;
    }

    int pivotIndex = s + count;
    swap(arr[pivotIndex], arr[s]);

    while (s < pivotIndex && e > pivotIndex)
    {
        while (arr[s] < pivot)
            s++;

        while (arr[e] > pivot)
            e--;

        if (s < pivotIndex && e > pivotIndex)
            swap(arr[s], arr[e]);
    }

    return pivotIndex;
}

void quickSort(int *arr, int s, int e)
{
    if (s >= e)
        return;

    int p = partition(arr, s, e);

    quickSort(arr, s, p - 1);
    quickSort(arr, p + 1, e);
}

int main()
{
    int arr[] = {4, 2, 1, 3, 5, 9, 6};
    int n = sizeof(arr) / sizeof(arr[0]);

    quickSort(arr, 0, n - 1);

    for (int i : arr)
        cout << i << " ";

    return 0;
}